[BUG] volume-backup-tester.py leaking backup and volume resources

I run the test against a cloud that's missing `cinder-backup`. The test creates a backup resource, which is in permanent error state because of the missing service, and I cannot delete it. This might be called a cinder bug, but the test triggered it. We cannot use a test that leaks resources.

What's more: the **test doesn't delete the volume either** because of the uncaught timeout exception...

My shell journey (abridged):

```shell
$ Tests/iaas/volume-backup/volume-backup-tester.py --os-cloud poc-wgcloud --debug

Performing cleanup for resources with the 'scs-test-' prefix ...
Creating volume ...
↳ waiting for volume with ID 'ffc4136f-df3f-4d00-8ad7-78a0dd2c04ad' to reach status 'available' ...
Create empty volume: PASS
Creating backup from volume ...
Traceback (most recent call last):
  File "/home/matthiasbuechse/opensrc/scs/standards/Tests/iaas/volume-backup/volume-backup-tester.py", line 283, in <module>
    main()
  File "/home/matthiasbuechse/opensrc/scs/standards/Tests/iaas/volume-backup/volume-backup-tester.py", line 278, in main
    test_backup(conn, prefix=args.prefix, timeout=args.timeout)
  File "/home/matthiasbuechse/opensrc/scs/standards/Tests/iaas/volume-backup/volume-backup-tester.py", line 115, in test_backup
    assert seconds_waited < timeout, (
           ^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: Timeout reached while waiting for backup to reach status 'available' (backup id: 86cc7f49-e91d-47b6-9668-959759820136) after 60 seconds

$ OS_CLOUD=poc-wgcloud .cargo/bin/osc block-storage backup show 86cc7f49-e91d-47b6-9668-959759820136
┌───────────────────────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Attribute             ┆ Value                                                                                                                                                                                                                                                                                                      │
╞═══════════════════════╪════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╡
│ created_at            ┆ 2024-10-30T19:55:23.000000                                                                                                                                                                                                                                                                                 │
│ data_timestamp        ┆ 2024-10-30T19:55:23.000000                                                                                                                                                                                                                                                                                 │
│ fail_reason           ┆ Service not found for creating backup.                                                                                                                                                                                                                                                                     │
│ has_dependent_backups ┆ false                                                                                                                                                                                                                                                                                                      │
│ id                    ┆ 86cc7f49-e91d-47b6-9668-959759820136                                                                                                                                                                                                                                                                       │
│ is_incremental        ┆ false                                                                                                                                                                                                                                                                                                      │
│ links                 ┆ [{"href":"https://block-storage.l1a.cloudandheat.com/v3/8ef56c4ec0a046f98ec3d95c7a677a6e/backups/86cc7f49-e91d-47b6-9668-959759820136","rel":"self"},{"href":"https://block-storage.l1a.cloudandheat.com/8ef56c4ec0a046f98ec3d95c7a677a6e/backups/86cc7f49-e91d-47b6-9668-959759820136","rel":"bookmark"}] │
│ metadata              ┆ {}                                                                                                                                                                                                                                                                                                         │
│ name                  ┆ scs-test-volume-backup                                                                                                                                                                                                                                                                                     │
│ object_count          ┆ 0                                                                                                                                                                                                                                                                                                          │
│ size                  ┆ 1                                                                                                                                                                                                                                                                                                          │
│ status                ┆ error                                                                                                                                                                                                                                                                                                      │
│ updated_at            ┆ 2024-10-30T19:55:23.000000                                                                                                                                                                                                                                                                                 │
│ volume_id             ┆ ffc4136f-df3f-4d00-8ad7-78a0dd2c04ad                                                                                                                                                                                                                                                                       │
└───────────────────────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

$ OS_CLOUD=poc-wgcloud .cargo/bin/osc block-storage backup force-delete 86cc7f49-e91d-47b6-9668-959759820136
Error: 
   0: openstack server error:
      	uri: `https://block-storage.l1a.cloudandheat.com/v3/8ef56c4ec0a046f98ec3d95c7a677a6e/backups/86cc7f49-e91d-47b6-9668-959759820136/action?`
      	status: `403 Forbidden`
      	data: `{"forbidden":{"code":403,"message":"Policy doesn't allow volume_extension:backup_admin_actions:force_delete to be performed."}}`

Location:
   openstack_cli/src/bin/osc.rs:31
```

--- a/Tests/iaas/volume-backup/volume-backup-tester.py
+++ b/Tests/iaas/volume-backup/volume-backup-tester.py
@@ -222,6 +222,10 @@ def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
                 # its own in the meantime ignore it
                 continue
             except ConformanceTestException as e:
+                # This exception happens if the backup state does not reach any
+                # of the desired ones specified above. We do not need to set
+                # cleanup_was_successful to False here since any remaining ones
+                # will be caught in the next loop down below anyway.
                 logging.warning(str(e))
             else:
                 logging.info(f"↳ deleting volume backup '{backup.id}' ...")

--- a/Tests/iaas/volume-backup/volume-backup-tester.py
+++ b/Tests/iaas/volume-backup/volume-backup-tester.py
@@ -164,10 +164,10 @@ def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
     """
 
     def wait_for_resource(resource_type: str, resource_id: str,
-                          expected_status="available") -> None:
+                          expected_status=["available"]) -> None:
         seconds_waited = 0
         get_func = getattr(conn.block_storage, f"get_{resource_type}")
-        while get_func(resource_id).status != expected_status:
+        while get_func(resource_id).status not in expected_status:
             time.sleep(1.0)
             seconds_waited += 1
             assert seconds_waited < timeout, (
@@ -184,7 +184,7 @@ def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
     for backup in backups:
         if backup.name.startswith(prefix):
             try:
-                wait_for_resource("backup", backup.id)
+                wait_for_resource("backup", backup.id, ["available", "error"])
             except openstack.exceptions.ResourceNotFound:
                 # if the resource has vanished on
                 # its own in the meantime ignore it
@@ -209,7 +209,7 @@ def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
     for volume in volumes:
         if volume.name.startswith(prefix):
             try:
-                wait_for_resource("volume", volume.id)
+                wait_for_resource("volume", volume.id, ["available", "error"])
             except openstack.exceptions.ResourceNotFound:
                 # if the resource has vanished on
                 # its own in the meantime ignore it

--- a/Tests/iaas/volume-backup/volume-backup-tester.py
+++ b/Tests/iaas/volume-backup/volume-backup-tester.py
@@ -29,6 +29,21 @@ DEFAULT_PREFIX = "scs-test-"
 WAIT_TIMEOUT = 60
 
 
+class ConformanceTestException(Exception):
+    pass
+
+
+def ensure(condition: bool, error_message: str):
+    """
+    Custom replacement for the `assert` statement that is not removed by the
+    -O optimization parameter.
+    If the condition does not evaluate to `True`, a ConformanceTestException
+    will be raised containing the specified error_message string.
+    """
+    if not condition:
+        raise ConformanceTestException(error_message)
+
+
 def connect(cloud_name: str, password: typing.Optional[str] = None
             ) -> openstack.connection.Connection:
     """Create a connection to an OpenStack cloud
@@ -64,17 +79,20 @@ def test_backup(conn: openstack.connection.Connection,
     """
 
     # CREATE VOLUME
-    print("Creating volume ...")
+    volume_name = f"{prefix}volume"
+    print(f"Creating volume '{volume_name}' ...")
     volume = conn.block_storage.create_volume(
-        name=f"{prefix}volume",
+        name=volume_name,
         size=1
     )
-    assert volume is not None, (
-        "Initial volume creation failed"
+    ensure(
+        volume is not None,
+        f"Creation of initial volume '{volume_name}' failed"
     )
     volume_id = volume.id
-    assert conn.block_storage.get_volume(volume_id) is not None, (
-        "Retrieving initial volume by ID failed"
+    ensure(
+        conn.block_storage.get_volume(volume_id) is not None,
+        f"Retrieving initial volume by ID '{volume_id}' failed"
     )
 
     print(
@@ -85,7 +103,8 @@ def test_backup(conn: openstack.connection.Connection,
     while conn.block_storage.get_volume(volume_id).status != "available":
         time.sleep(1.0)
         seconds_waited += 1
-        assert seconds_waited < timeout, (
+        ensure(
+            seconds_waited < timeout,
             f"Timeout reached while waiting for volume to reach status "
             f"'available' (volume id: {volume_id}) after {seconds_waited} "
             f"seconds"
@@ -98,11 +117,13 @@ def test_backup(conn: openstack.connection.Connection,
         name=f"{prefix}volume-backup",
         volume_id=volume_id
     )
-    assert backup is not None, (
+    ensure(
+        backup is not None,
         "Backup creation failed"
     )
     backup_id = backup.id
-    assert conn.block_storage.get_backup(backup_id) is not None, (
+    ensure(
+        conn.block_storage.get_backup(backup_id) is not None,
         "Retrieving backup by ID failed"
     )
 
@@ -111,7 +132,8 @@ def test_backup(conn: openstack.connection.Connection,
     while conn.block_storage.get_backup(backup_id).status != "available":
         time.sleep(1.0)
         seconds_waited += 1
-        assert seconds_waited < timeout, (
+        ensure(
+            seconds_waited < timeout,
             f"Timeout reached while waiting for backup to reach status "
             f"'available' (backup id: {backup_id}) after {seconds_waited} "
             f"seconds"
@@ -119,8 +141,8 @@ def test_backup(conn: openstack.connection.Connection,
     print("Create backup from volume: PASS")
 
     # RESTORE BACKUP
-    print("Restoring backup to volume ...")
     restored_volume_name = f"{prefix}restored-backup"
+    print(f"Restoring backup to volume '{restored_volume_name}' ...")
     conn.block_storage.restore_backup(
         backup_id,
         name=restored_volume_name
@@ -134,7 +156,8 @@ def test_backup(conn: openstack.connection.Connection,
     while conn.block_storage.find_volume(restored_volume_name) is None:
         time.sleep(1.0)
         seconds_waited += 1
-        assert seconds_waited < timeout, (
+        ensure(
+            seconds_waited < timeout,
             f"Timeout reached while waiting for restored volume to be created "
             f"(volume name: {restored_volume_name}) after {seconds_waited} "
             f"seconds"
@@ -148,7 +171,8 @@ def test_backup(conn: openstack.connection.Connection,
     while conn.block_storage.get_volume(volume_id).status != "available":
         time.sleep(1.0)
         seconds_waited += 1
-        assert seconds_waited < timeout, (
+        ensure(
+            seconds_waited < timeout,
             f"Timeout reached while waiting for restored volume reach status "
             f"'available' (volume id: {volume_id}) after {seconds_waited} "
             f"seconds"
@@ -170,9 +194,10 @@ def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
         while get_func(resource_id).status not in expected_status:
             time.sleep(1.0)
             seconds_waited += 1
-            assert seconds_waited < timeout, (
+            ensure(
+                seconds_waited < timeout,
                 f"Timeout reached while waiting for {resource_type} during "
-                f"cleanup to be in status '{expected_status}' "
+                f"cleanup to be in status {expected_status} "
                 f"({resource_type} id: {resource_id}) after {seconds_waited} "
                 f"seconds"
             )
@@ -200,7 +225,8 @@ def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
     ) > 0:
         time.sleep(1.0)
         seconds_waited += 1
-        assert seconds_waited < timeout, (
+        ensure(
+            seconds_waited < timeout,
             f"Timeout reached while waiting for all backups with prefix "
             f"'{prefix}' to finish deletion"
         )
@@ -262,10 +288,11 @@ def main():
     cloud = os.environ.get("OS_CLOUD", None)
     if args.os_cloud:
         cloud = args.os_cloud
-    assert cloud, (
-        "You need to have the OS_CLOUD environment variable set to your "
-        "cloud name or pass it via --os-cloud"
-    )
+    if not cloud:
+        raise Exception(
+            "You need to have the OS_CLOUD environment variable set to your "
+            "cloud name or pass it via --os-cloud"
+        )
     conn = connect(
         cloud,
         password=getpass.getpass("Enter password: ") if args.ask else None

--- a/Tests/iaas/volume-backup/volume-backup-tester.py
+++ b/Tests/iaas/volume-backup/volume-backup-tester.py
@@ -301,8 +301,10 @@ def main():
         cleanup(conn, prefix=args.prefix, timeout=args.timeout)
     else:
         cleanup(conn, prefix=args.prefix, timeout=args.timeout)
-        test_backup(conn, prefix=args.prefix, timeout=args.timeout)
-        cleanup(conn, prefix=args.prefix, timeout=args.timeout)
+        try:
+            test_backup(conn, prefix=args.prefix, timeout=args.timeout)
+        finally:
+            cleanup(conn, prefix=args.prefix, timeout=args.timeout)
 
 
 if __name__ == "__main__":

--- a/Tests/iaas/volume-backup/volume-backup-tester.py
+++ b/Tests/iaas/volume-backup/volume-backup-tester.py
@@ -190,7 +190,7 @@ def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
     """
 
     def wait_for_resource(resource_type: str, resource_id: str,
-                          expected_status=["available"]) -> None:
+                          expected_status=("available", )) -> None:
         seconds_waited = 0
         get_func = getattr(conn.block_storage, f"get_{resource_type}")
         while get_func(resource_id).status not in expected_status:
@@ -212,7 +212,7 @@ def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
     for backup in backups:
         if backup.name.startswith(prefix):
             try:
-                wait_for_resource("backup", backup.id, ["available", "error"])
+                wait_for_resource("backup", backup.id, expected_status=("available", "error"))
             except openstack.exceptions.ResourceNotFound:
                 # if the resource has vanished on
                 # its own in the meantime ignore it
@@ -247,7 +247,7 @@ def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
     for volume in volumes:
         if volume.name.startswith(prefix):
             try:
-                wait_for_resource("volume", volume.id, ["available", "error"])
+                wait_for_resource("volume", volume.id, expected_status=("available", "error"))
             except openstack.exceptions.ResourceNotFound:
                 # if the resource has vanished on
                 # its own in the meantime ignore it

--- a/Tests/iaas/volume-backup/volume-backup-tester.py
+++ b/Tests/iaas/volume-backup/volume-backup-tester.py
@@ -332,22 +332,22 @@ def main():
         cloud,
         password=getpass.getpass("Enter password: ") if args.ask else None
     )
+
+    if not cleanup(conn, prefix=args.prefix, timeout=args.timeout):
+        raise Exception(
+            f"Cleanup was not successful, there may be leftover resources "
+            f"with the '{args.prefix}' prefix"
+        )
     if args.cleanup_only:
+        return
+    try:
+        test_backup(conn, prefix=args.prefix, timeout=args.timeout)
+    finally:
         if not cleanup(conn, prefix=args.prefix, timeout=args.timeout):
-            raise Exception(
-                f"Cleanup was not successful, there may be leftover resources "
-                f"with the '{args.prefix}' prefix"
+            logging.info(
+                f"There may be leftover resources with the "
+                f"'{args.prefix}' prefix that could not be cleaned up!"
             )
-    else:
-        cleanup(conn, prefix=args.prefix, timeout=args.timeout)
-        try:
-            test_backup(conn, prefix=args.prefix, timeout=args.timeout)
-        finally:
-            if not cleanup(conn, prefix=args.prefix, timeout=args.timeout):
-                logging.info(
-                    f"There may be leftover resources with the "
-                    f"'{args.prefix}' prefix that could not be cleaned up!"
-                )
 
 
 if __name__ == "__main__":

--- a/Tests/iaas/volume-backup/volume-backup-tester.py
+++ b/Tests/iaas/volume-backup/volume-backup-tester.py
@@ -181,10 +181,12 @@ def test_backup(conn: openstack.connection.Connection,
 
 
 def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
-            timeout=WAIT_TIMEOUT):
+            timeout=WAIT_TIMEOUT) -> bool:
     """
     Looks up volume and volume backup resources matching the given prefix and
     deletes them.
+    Returns False if there were any errors during cleanup which might leave
+    resources behind. Otherwise returns True to indicate cleanup success.
     """
 
     def wait_for_resource(resource_type: str, resource_id: str,
@@ -205,6 +207,7 @@ def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
     print(f"\nPerforming cleanup for resources with the "
           f"'{prefix}' prefix ...")
 
+    cleanup_was_successful = True
     backups = conn.block_storage.backups()
     for backup in backups:
         if backup.name.startswith(prefix):
@@ -214,8 +217,11 @@ def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
                 # if the resource has vanished on
                 # its own in the meantime ignore it
                 continue
-            print(f"↳ deleting volume backup '{backup.id}' ...")
-            conn.block_storage.delete_backup(backup.id)
+            except ConformanceTestException as e:
+                print("WARNING:", str(e))
+            else:
+                print(f"↳ deleting volume backup '{backup.id}' ...")
+                conn.block_storage.delete_backup(backup.id)
 
     # wait for all backups to be cleaned up before attempting to remove volumes
     seconds_waited = 0
@@ -225,11 +231,17 @@ def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
     ) > 0:
         time.sleep(1.0)
         seconds_waited += 1
-        ensure(
-            seconds_waited < timeout,
-            f"Timeout reached while waiting for all backups with prefix "
-            f"'{prefix}' to finish deletion"
-        )
+        try:
+            ensure(
+                seconds_waited < timeout,
+                f"Timeout reached while waiting for all backups with prefix "
+                f"'{prefix}' to finish deletion during cleanup after "
+                f"{seconds_waited} seconds"
+            )
+        except ConformanceTestException as e:
+            print("WARNING:", str(e))
+            cleanup_was_successful = False
+            break
 
     volumes = conn.block_storage.volumes()
     for volume in volumes:
@@ -240,8 +252,14 @@ def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
                 # if the resource has vanished on
                 # its own in the meantime ignore it
                 continue
-            print(f"↳ deleting volume '{volume.id}' ...")
-            conn.block_storage.delete_volume(volume.id)
+            except ConformanceTestException as e:
+                print("WARNING:", str(e))
+                cleanup_was_successful = False
+            else:
+                print(f"↳ deleting volume '{volume.id}' ...")
+                conn.block_storage.delete_volume(volume.id)
+
+    return cleanup_was_successful
 
 
 def main():
@@ -298,13 +316,21 @@ def main():
         password=getpass.getpass("Enter password: ") if args.ask else None
     )
     if args.cleanup_only:
-        cleanup(conn, prefix=args.prefix, timeout=args.timeout)
+        if not cleanup(conn, prefix=args.prefix, timeout=args.timeout):
+            raise Exception(
+                f"Cleanup was not successful, there may be leftover resources "
+                f"with the '{args.prefix}' prefix"
+            )
     else:
         cleanup(conn, prefix=args.prefix, timeout=args.timeout)
         try:
             test_backup(conn, prefix=args.prefix, timeout=args.timeout)
         finally:
-            cleanup(conn, prefix=args.prefix, timeout=args.timeout)
+            if not cleanup(conn, prefix=args.prefix, timeout=args.timeout):
+                print(
+                    f"There may be leftover resources with the "
+                    f"'{args.prefix}' prefix that could not be cleaned up!"
+                )
 
 
 if __name__ == "__main__":

--- a/Tests/iaas/volume-backup/volume-backup-tester.py
+++ b/Tests/iaas/volume-backup/volume-backup-tester.py
@@ -231,16 +231,14 @@ def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
     ) > 0:
         time.sleep(1.0)
         seconds_waited += 1
-        try:
-            ensure(
-                seconds_waited < timeout,
-                f"Timeout reached while waiting for all backups with prefix "
+        if seconds_waited >= timeout:
+            cleanup_was_successful = False
+            print(
+                "WARNING:"
+                "Timeout reached while waiting for all backups with prefix "
                 f"'{prefix}' to finish deletion during cleanup after "
                 f"{seconds_waited} seconds"
             )
-        except ConformanceTestException as e:
-            print("WARNING:", str(e))
-            cleanup_was_successful = False
             break
 
     volumes = conn.block_storage.volumes()

--- a/Tests/iaas/volume-backup/volume-backup-tester.py
+++ b/Tests/iaas/volume-backup/volume-backup-tester.py
@@ -17,6 +17,7 @@ import getpass
 import os
 import time
 import typing
+import logging
 
 import openstack
 
@@ -80,7 +81,7 @@ def test_backup(conn: openstack.connection.Connection,
 
     # CREATE VOLUME
     volume_name = f"{prefix}volume"
-    print(f"Creating volume '{volume_name}' ...")
+    logging.info(f"Creating volume '{volume_name}' ...")
     volume = conn.block_storage.create_volume(
         name=volume_name,
         size=1
@@ -95,7 +96,7 @@ def test_backup(conn: openstack.connection.Connection,
         f"Retrieving initial volume by ID '{volume_id}' failed"
     )
 
-    print(
+    logging.info(
         f"↳ waiting for volume with ID '{volume_id}' to reach status "
         f"'available' ..."
     )
@@ -109,10 +110,10 @@ def test_backup(conn: openstack.connection.Connection,
             f"'available' (volume id: {volume_id}) after {seconds_waited} "
             f"seconds"
         )
-    print("Create empty volume: PASS")
+    logging.info("Create empty volume: PASS")
 
     # CREATE BACKUP
-    print("Creating backup from volume ...")
+    logging.info("Creating backup from volume ...")
     backup = conn.block_storage.create_backup(
         name=f"{prefix}volume-backup",
         volume_id=volume_id
@@ -127,7 +128,7 @@ def test_backup(conn: openstack.connection.Connection,
         "Retrieving backup by ID failed"
     )
 
-    print(f"↳ waiting for backup '{backup_id}' to become available ...")
+    logging.info(f"↳ waiting for backup '{backup_id}' to become available ...")
     seconds_waited = 0
     while conn.block_storage.get_backup(backup_id).status != "available":
         time.sleep(1.0)
@@ -138,17 +139,17 @@ def test_backup(conn: openstack.connection.Connection,
             f"'available' (backup id: {backup_id}) after {seconds_waited} "
             f"seconds"
         )
-    print("Create backup from volume: PASS")
+    logging.info("Create backup from volume: PASS")
 
     # RESTORE BACKUP
     restored_volume_name = f"{prefix}restored-backup"
-    print(f"Restoring backup to volume '{restored_volume_name}' ...")
+    logging.info(f"Restoring backup to volume '{restored_volume_name}' ...")
     conn.block_storage.restore_backup(
         backup_id,
         name=restored_volume_name
     )
 
-    print(
+    logging.info(
         f"↳ waiting for restoration target volume '{restored_volume_name}' "
         f"to be created ..."
     )
@@ -163,7 +164,7 @@ def test_backup(conn: openstack.connection.Connection,
             f"seconds"
         )
     # wait for the volume restoration to finish
-    print(
+    logging.info(
         f"↳ waiting for restoration target volume '{restored_volume_name}' "
         f"to reach 'available' status ..."
     )
@@ -177,7 +178,7 @@ def test_backup(conn: openstack.connection.Connection,
             f"'available' (volume id: {volume_id}) after {seconds_waited} "
             f"seconds"
         )
-    print("Restore volume from backup: PASS")
+    logging.info("Restore volume from backup: PASS")
 
 
 def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
@@ -204,8 +205,8 @@ def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
                 f"seconds"
             )
 
-    print(f"\nPerforming cleanup for resources with the "
-          f"'{prefix}' prefix ...")
+    logging.info(f"Performing cleanup for resources with the "
+                 f"'{prefix}' prefix ...")
 
     cleanup_was_successful = True
     backups = conn.block_storage.backups()
@@ -218,9 +219,9 @@ def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
                 # its own in the meantime ignore it
                 continue
             except ConformanceTestException as e:
-                print("WARNING:", str(e))
+                logging.warning(str(e))
             else:
-                print(f"↳ deleting volume backup '{backup.id}' ...")
+                logging.info(f"↳ deleting volume backup '{backup.id}' ...")
                 conn.block_storage.delete_backup(backup.id)
 
     # wait for all backups to be cleaned up before attempting to remove volumes
@@ -233,9 +234,8 @@ def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
         seconds_waited += 1
         if seconds_waited >= timeout:
             cleanup_was_successful = False
-            print(
-                "WARNING:"
-                "Timeout reached while waiting for all backups with prefix "
+            logging.warning(
+                f"Timeout reached while waiting for all backups with prefix "
                 f"'{prefix}' to finish deletion during cleanup after "
                 f"{seconds_waited} seconds"
             )
@@ -251,10 +251,10 @@ def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
                 # its own in the meantime ignore it
                 continue
             except ConformanceTestException as e:
-                print("WARNING:", str(e))
+                logging.warning(str(e))
                 cleanup_was_successful = False
             else:
-                print(f"↳ deleting volume '{volume.id}' ...")
+                logging.info(f"↳ deleting volume '{volume.id}' ...")
                 conn.block_storage.delete_volume(volume.id)
 
     return cleanup_was_successful
@@ -299,6 +299,10 @@ def main():
     )
     args = parser.parse_args()
     openstack.enable_logging(debug=args.debug)
+    logging.basicConfig(
+        format="%(levelname)s: %(message)s",
+        level=logging.DEBUG if args.debug else logging.INFO,
+    )
 
     # parse cloud name for lookup in clouds.yaml
     cloud = os.environ.get("OS_CLOUD", None)
@@ -325,7 +329,7 @@ def main():
             test_backup(conn, prefix=args.prefix, timeout=args.timeout)
         finally:
             if not cleanup(conn, prefix=args.prefix, timeout=args.timeout):
-                print(
+                logging.info(
                     f"There may be leftover resources with the "
                     f"'{args.prefix}' prefix that could not be cleaned up!"
                 )

--- a/Tests/iaas/volume-backup/volume-backup-tester.py
+++ b/Tests/iaas/volume-backup/volume-backup-tester.py
@@ -213,7 +213,10 @@ def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
     for backup in backups:
         if backup.name.startswith(prefix):
             try:
-                wait_for_resource("backup", backup.id, expected_status=("available", "error"))
+                wait_for_resource(
+                    "backup", backup.id,
+                    expected_status=("available", "error")
+                )
             except openstack.exceptions.ResourceNotFound:
                 # if the resource has vanished on
                 # its own in the meantime ignore it
@@ -222,7 +225,15 @@ def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
                 logging.warning(str(e))
             else:
                 logging.info(f"↳ deleting volume backup '{backup.id}' ...")
-                conn.block_storage.delete_backup(backup.id)
+                # Setting ignore_missing to False here will make an exception
+                # bubble up in case the cinder-backup service is not present.
+                # Since we already catch ResourceNotFound for the backup above,
+                # the absence of the cinder-backup service is the only
+                # NotFoundException that is left to be thrown here.
+                # We treat this as a fatal due to the cinder-backup service
+                # being mandatory.
+                conn.block_storage.delete_backup(
+                    backup.id, ignore_missing=False)
 
     # wait for all backups to be cleaned up before attempting to remove volumes
     seconds_waited = 0

